[CPF-157] Medusa

jlhughes opened the pull request:
Notes on use are on the matching nanopay branch PR

Non-Medusa Changes
* boot script "main" moved into own journal so it can be instance specific. 
...

jlhughes commented on src/foam/box/MessageReplyBox.js:
```diff
@@ -0,0 +1,35 @@
+/**
```
To be removed.

jlhughes commented on src/foam/box/SessionServerBox.java:
```diff
@@ -42,29 +45,82 @@ public SessionServerBox(X x, Box delegate, boolean authenticate) {
+                if ( sessionID != null ) {
+                  session = (Session) sessionDAO.find(sessionID);
+                  if ( session == null ) {
+                    // test and use non-clustered medusa sessions
```
NOTE: logging in with Load-Balancers has not been tested yet.

jlhughes replied:
admin login is only supported on explicit instance.

jlhughes commented on src/foam/box/socket/BoxHolder.js:
```diff
@@ -0,0 +1,22 @@
+/**
```
Used by SocketConnectionBox

nanoscott commented on src/foam/box/SessionServerBox.java:
```diff
@@ -42,29 +45,82 @@ public SessionServerBox(X x, Box delegate, boolean authenticate) {
+                    // test and use non-clustered medusa sessions
+                    DAO internalSessionDAO = (DAO) getX().get("internalSessionDAO");
+                    if ( internalSessionDAO != null ) {
+                      session = (Session) internalSessionDAO.find(sessionID);
```
question: how big of a security hole is this? if an external actor got ahold of these internal session IDs, what permissions would they have?

follow-up question: if they could get access to these internal sesssion IDs, then it probably means they already have access to the system in some deep way .. so does the above question matter? Probably only if the internal session IDs are constant or otherwise discoverable.

jlhughes replied:
In general, yes, more discussion needed here.  
The above is just so I can login before the cluster is running.
The real session/token issue at the socket connections between mediators and nodes. 
And this where we need to make use of non-static, rolling, hsm, ....  identifiers. 
For testing they all have known GUIDs.

nanoscott commented on src/foam/box/socket/SocketConnectionBox.js:
```diff
@@ -0,0 +1,272 @@
+      javaCode: `
+      PM pm = PM.create(getX(), this.getClass().getSimpleName(), getHost()+":"+getPort()+":send");
+      Box replyBox = (Box) msg.getAttributes().get("replyBox");
+      if ( replyBox != null ) {
```
why does it seem like the code below should only be executed if replyBox == null, not replyBox != null? what is happening when the replyBox is null?

jlhughes replied:
When replyBox is null, it's like send and forget. 
When replyBox is not null, we need to keep track of it, as the SocketConnectionBox is replacing the ReplyBox with its own.  When the response is received, we lookup the original reply box and call it.

nanoscott commented on src/foam/box/socket/SocketServerProcessor.java:
```diff
@@ -0,0 +1,153 @@
+          byte[] bytes = new byte[length];
+          StringBuilder data = new StringBuilder();
+          int total = 0;
+          while ( true ) {
```
is this the same code as in the SocketConnectionBox?

jlhughes replied:
yes, the connection box is waiting on a reply just like the server for a request. 
We could abstract into a InputSocketReader or something.

jlhughes commented on src/foam/nanos/crunch/cronjobs:
```diff
@@ -16,7 +16,7 @@ p({
     print("done");
   """,
-  "enabled":true
+  "enabled":false
```
Disable during medusa testing as we have a midnight cron bug - run repeated in the midnight hour.

nanoscott reviewed (approved):
LGTM .. there are some TODOs in the code and some minor formatting errors, but of all the code it seems sound. There was some stuff that I didn't follow, but I went in depth in the consensus algorithm, the voting algorithm and the ClusterConfigSupport code. I reviewed the Script/ScriptEvent updates as well.

jlhughes committed:
comment out really verbose debug log

jlhughes committed:
comment out really verbose debug log

jlhughes committed:
cleanup logger output

lchanmann commented on src/foam/nanos/alarming/AlarmLoggingDAO.js:
```diff
@@ -0,0 +1,45 @@
+      Alarm alarm = (Alarm) getDelegate().put_(x, obj);
+      Logger logger = (Logger) x.get("logger");
+      switch ( alarm.getSeverity().getName() ) {
+        case "DEBUG":
```
Can the switch case be on the enum instead of the name?
```suggestion
      switch ( alarm.getSeverity() ) {
        case DEBUG:
```

jlhughes replied:
yes

lchanmann commented on src/foam/nanos/auth/AuthorizationDAO.js:
```diff
@@ -59,7 +58,7 @@ foam.CLASS({
 
     Object id = obj.getProperty("id");
-    FObject oldObj = getDelegate().inX(x).find(id);
+    FObject oldObj = id == null ? null : getDelegate().inX(x).find(id);
```
Just curious, do we check for id = null to avoid going through `find`? for a model using Long for the id property it won't be null.

jlhughes replied:
```suggestion
    FObject oldObj = getDelegate().inX(x).find(id);
```

lchanmann commented on src/foam/nanos/medusa/ClusterClientDAO.js:
```diff
@@ -0,0 +1,247 @@
+      if ( getNSpec() != null ) {
+        return getNSpec().getName();
+      }
+      return serviceName_;
```
String property factory in FOAM falls back to empty string, `return serviceName_` could make it null, is it intended?

jlhughes replied:
serviceName is no longer required. It was earlier on.  ClusterClientDAOs are created by MedusaAdapterDAOs and some time ago they didn't have NSpec.  And I was flipping between using an NSpec directly, or NSpec.name.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
add JSESSIONID for sticket sessions

jlhughes committed:
set Message.exception on reply

jlhughes committed:
refactor to extend foam.core.FOAMException

jlhughes committed:
move order to served DAO and correct comparator so displayed in descending order by created date

jlhughes committed:
bareUserCapabilityJunction - explicitly add mdao, and add indexes.

jlhughes committed:
correct typo in comment

jlhughes committed:
Update for PurgePromotionAgent

jlhughes committed:
disable very verbose debugging

lchanmann commented on src/foam/nanos/medusa/ClusterConfigSupport.js:
```diff
@@ -0,0 +1,1051 @@
+          ClusterConfig cfg = configs.get(0);
+          getLogger().info("nextZone", "configs", configs.size(), "selected", cfg.getId(), cfg.getZone(), cfg.getIsPrimary(), cfg.getPingTime());
+          for ( ClusterConfig c : configs ) {
+            getLogger().info("nextZone", "other", cfg.getId(), cfg.getZone(), cfg.getIsPrimary(), cfg.getPingTime());
```
wondering should it log `c` instead of `cfg`?
```suggestion
          for ( ClusterConfig c : configs ) {
            getLogger().info("nextZone", "other", c.getId(), c.getZone(), c.getIsPrimary(), c.getPingTime());
```

jlhughes committed:
cluster email - for testing

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
update expected hash to reflect change in 'includeInDigest' properties. Previously firstname, lastname were not included.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

lchanmann commented on src/foam/nanos/medusa/ClusterConfigMonitor.js:
```diff
@@ -0,0 +1,210 @@
+
+        // update ReplayingInfo for self.
+        DAO dao = (DAO) x.get("clusterConfigDAO");
+        dao.find_(x, config.getId());
```
```suggestion
        DAO dao = (DAO) x.get("clusterConfigDAO");
        config = (ClusterConfig) dao.find_(x, config.getId());
```
Might not need to reload the config in case the id and realm were unchanging.

jlhughes replied:
See ClusterConfigReplayingInfoDAO - the find on self, is just to invoke this decorator which updates the config for network graph.

lchanmann commented on src/foam/nanos/medusa/ClusterConfigMonitor.js:
```diff
@@ -0,0 +1,210 @@
+  package: 'foam.nanos.medusa',
+  name: 'ClusterConfigMonitor',
+
+  documentation: 'NOTE: do not start with cronjob. This process starts the ClusterConfigPingSing which polls the Mediators and Nodes and will initiate Replay, and Elections.',
```
```suggestion
  documentation: 'NOTE: do not start with cronjob. This process starts the ClusterConfigMonitorAgents which polls the Mediators and Nodes and will initiate Replay, and Elections.',
```

lchanmann commented on src/foam/nanos/medusa/ClusterConfigReplayDAO.js:
```diff
@@ -0,0 +1,188 @@
+          replaying.setStartTime(new java.util.Date());
+
+          DAO dao = (DAO) x.get("medusaNodeDAO");
+          Min min = (Min) MIN(MedusaEntry.INDEX);
```
min is unused?

jlhughes replied:
Yes.  The incomplete intent is make a more focused replay request to the nodes. Presently max is ignored as well, and the nodes just replay everything.

jlhughes committed:
Update src/foam/nanos/medusa/ClusterConfigMonitor.js

Co-authored-by: Chanmann Lim <chanmann@nanopay.net>

jlhughes committed:
Update src/foam/nanos/medusa/ClusterConfigMonitor.js

Co-authored-by: Chanmann Lim <chanmann@nanopay.net>

jlhughes committed:
Update src/foam/nanos/medusa/ClusterConfigSupport.js

Co-authored-by: Chanmann Lim <chanmann@nanopay.net>

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'CPF-157/Medusa' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
generate NOC notification on admin login (pass and fail), also log subject (user, agent) junction missing.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
use ToastMessage for NotificationTemplate email messages

jlhughes committed:
syntax error

jlhughes committed:
compile before commit

jlhughes committed:
temporarily remove CapabilityAvailabilityDAO from capabilityDAO else no capabilities are accessible to test the system

jlhughes committed:
Update src/foam/nanos/auth/AuthorizationDAO.js

jlhughes committed:
do not generate NOC Notification for admin login, clustering may not be available and it will retry forever.

jlhughes committed:
copy clusterable

jlhughes committed:
Merge branch 'CPF-157/Medusa' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
set clusterable(false) on login Notifications.

jlhughes committed:
implement Clusterable

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
re-enable CapabilityAvailabilityDAO

jlhughes committed:
fix use of userX

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
remove log call

jlhughes committed:
refactor to handle sessionID from HttpRequest and Box message

jlhughes committed:
remove space which generates parsing warning

jlhughes committed:
comment out very verbose debug

jlhughes committed:
cleanup

jlhughes committed:
Return false when realUser is null

jlhughes committed:
remove Notification - incompatible with admin login before clustering ready

jlhughes committed:
remove unused import

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
allow for smaller node bucket groups

jlhughes committed:
add CIDR Whitelist support

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Remove JSESSIONID for sticky sessions

jlhughes committed:
remove an unnecessary inX

jlhughes committed:
Move SequenceNumberDAO and GUIDDAO before MedusaAdapter, else all mediators will be picking the same next id

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Handle just one zone

jlhughes committed:
replace hard-coded 'nanopay' realm with realm from clusterconfig

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Re-add sessionID property for Java outbound box use

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

nanoscott commented on src/foam/box/socket/SocketConnectionBox.js:
```diff
@@ -0,0 +1,314 @@
+             ! socket.isConnected() ) {
+          throw new SocketException("Socket not connected.");
+        }
+        // omLogger.log(this.getClass().getSimpleName(), getId(), "pending");
```
several log messages are commented out in this class .. either remove them or uncomment them

nanoscott commented on src/foam/box/socket/SocketConnectionBoxManager.js:
```diff
@@ -0,0 +1,218 @@
+      ],
+      javaCode: `
+      if ( box != null ) {
+//        getBoxes().remove(makeKey(box.getHost(), box.getPort()));
```
remove dead code

nanoscott commented on src/foam/crypto/hash/HashableTest.js:
```diff
@@ -28,15 +28,15 @@ foam.CLASS({
         // non-chained hash with correct digest
         Hashable_HashWithValidAlgorithm(input, "MD5",
-            "47f68c52aaae1c2cdad696c59a248752");
+            "d41d8cd98f00b204e9800998ecf8427e");
```
I just had to change these too, so they may need to be updated again.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

nanoscott commented on src/foam/dao/EasyDAO.js:
```diff
@@ -606,8 +612,17 @@ foam.CLASS({
+      name: 'cluster',
+      class: 'Boolean',
+      javaFactory: `
+    return "true".equals(System.getProperty("CLUSTER"));
```
```suggestion
    return SafetyUtil.equals("true", System.getProperty("CLUSTER"));
```

nanoscott commented on src/foam/dao/EasyDAO.js:
```diff
@@ -208,6 +233,12 @@ foam.CLASS({
+        // DelegateFactory calls from subsequent DAOs which may
+        // have init_ methods which in turn call getDelegate().
+        delegateIsSet_ = true;
+        delegate_ = new ProxyDAO.Builder(getX()).setDelegate(delegate).build();
```
this seems a bit concerning that the code within the factory calls getDelegate on EasyDAO before the factory is finished ... is the reentrance unavoidable?

jlhughes replied:
The reentrant call would need to receive something else we'll deadlock.  Not sure how to avoid this.

nanoscott commented on src/foam/nanos/auth/CapabilityAuthService.js:
```diff
@@ -144,6 +144,9 @@ foam.CLASS({
         if ( x.get(Session.class) == null ) return false;
         if ( user == null || ! user.getEnabled() ) return false;
         User realUser = ((Subject) x.get("subject")).getRealUser();
+        if ( realUser == null ) {
```
might be a merging artifact .. doesn't seem like a clustering change to me.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
revert User hashes, as this branch explicitly sets includeInDigest and storageOptional

jlhughes committed:
replace table columns description with id as description not always set

jlhughes committed:
set storageOptional on lastModified

jlhughes committed:
Order by Enabled and Last_Run.  Rename to .jrl. Add 'id' to table columns as description not always filled in.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
revert to medusa hashes

jlhughes committed:
add medusa api user

jlhughes committed:
rename sha256 directory to files

jlhughes committed:
cleanup journal type selection logic

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
cleanup, address PR comments

jlhughes committed:
add hostname to track origin of exception in components such as medusa

jlhughes committed:
change toString format, output hostname first

jlhughes committed:
add additional AuthenticationExceptions such as AccessDenied to better control which scenarios will lock out an account - originally intended to protect admin from a DOS attack when CIDR restrictions are enabled.

jlhughes committed:
add AccessDeniedException and InvalidPasswordException

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
remove exception generation in debug

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
correct merge

jlhughes committed:
Update for JSONFormatter changes

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
add notify flag to request explicit notification on script execution

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Test Electoral Service status for load balance 'up'

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
correct Journal setup when running in cluster

jlhughes committed:
resolve syntax error and fix test on ElectoraService

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
cleanup/review node quorum logic and bucket assignment logic

jlhughes committed:
Report but don't throw exception if no Theme found

jlhughes committed:
Gap testing - use internalMedusaDAO so not to block on ReplayingDAO

jlhughes committed:
Change includeInDigest to default to true

jlhughes committed:
Review includeInDigest

jlhughes committed:
syntax error

jlhughes committed:
resolve client side files complaint

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Update ClusterConfig.isPrimary when quorum lost

jlhughes committed:
remove comemnt

jlhughes committed:
Merge branch 'authenticated-dao-fix' of github.com:nanoscott/foam2 into CPF-157/Medusa

jlhughes committed:
add SocketServer to avoid DIG issues listing nspecs

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Add non-zero purgeTime

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Use logged-in user session in SessionClientBox calls so remote is acting as user

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
update digests to match medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
fix bad merged

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
support all medusa types

jlhughes committed:
Refactor to simply use explicit url configuration

jlhughes committed:
Use zone 1 circle if only one zone to display

jlhughes committed:
Avoid vote on mediator ONLINE if quorum already achieved and IN-SESSION

jlhughes committed:
support removeAll

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
increase table width

jlhughes committed:
Acquire appConfig via group to ensure url is correct

jlhughes committed:
Merge branch 'NP-3116/SmarterCompletionIndicator' of github.com:KernelDeimos/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'NP-3053/FixUpdateUserJunction' of github.com:KernelDeimos/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
correct strange merge

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Support clustering the nSpecDAO

jlhughes committed:
documentation

jlhughes committed:
Add logic to bring n+1 mediator IN_SESSION when cluster already has quorum when it comes ONLINE

jlhughes committed:
Clone before update and put

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Reconfigure Alarm, Notification, Logging interaction

* Logging no longer generates Notifications for WARN, ERROR.
* Alarm generates Notification on status change for WARN, ERROR level Alarms.
* Add Alarm 'alarm' email template for Notifications.
* Notification resend action now reports failures.
* Add AlarmId convenience constructor taking just Alarm name.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
fix syntax error

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
correct bad merge

jlhughes committed:
Non-Active Region Update

jlhughes committed:
Standby (Non-Active) Region synchronization working.
Haven't tested fail-over yet.

jlhughes committed:
Update CViews for Multiple regions

jlhughes committed:
show Alarm count on ReplayingInfoDetailCView

jlhughes committed:
generate toast message when trying to run cron jobs from non-primary

jlhughes committed:
removed incorrect find

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Add toast notifications on script execution. Also convert text to MESSAGES

jlhughes committed:
correct MESSAGE use in choices and tableCellFormatter

jlhughes committed:
add Since time

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
ServiceProviderAwareDAO DAO operations were not refactored for move of spid to models

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
DAO and Test working

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Make UserCapabilityJunction LIfecycleAware to support remove/delete of UCJs.

Fixing ServiceProviderAwareDAO resulted User/Contact/Business remove/removeAll test
failing on the cascade remove (mark delete) on associated UCJs.  For some reason
this cascade remove was not occuring before.

The LifecycleAwareDAO will mark removed UCJs as LifecycleState.DELETED.
The remainder of CRUNCH still needs to be updated to handled these deleted ucjs.

NP-3217 #time 2d

jlhughes committed:
cleanup

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'CPF-157/Medusa' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Remove userCapabilityJunctionHistoryDAO from nspec, so dynamic api tests don't pick it up.

jlhughes committed:
add additional loop time to RulerDAOTest to wait on async to complete

jlhughes committed:
Merge branch 'NP-3217/ServiceProviderAwareDAOBroken' into CPF-157/Medusa

jlhughes committed:
Alarm on missing MDAO rather than fail the system.

jlhughes committed:
update documentation

jlhughes committed:
Remove deleted classes

jlhughes committed:
Manually copy from foam master to recover from a bad merge at some point

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

kgrgreer commented on src/foam/box/socket/SocketConnectionBoxManager.js:
```diff
@@ -0,0 +1,218 @@
+          SocketAddress address = new InetSocketAddress(host, port);
+          socket.connect(address, getConnectTimeout());
+          box = new SocketConnectionBox(x, key, socket, host, port);
+          add(box);
```
The 'boxes' collection is synchronized but there's no sync at this level to prevent two from being created. Is that going to be a problem?

jlhughes replied:
The model method is synchronized: 
`    {
      name: 'get',
      synchronized: true,
      type: 'foam.box.Box',
`

kgrgreer commented on src/foam/box/socket/SocketServer.js:
```diff
@@ -0,0 +1,111 @@
+  package: 'foam.box.socket',
+  name: 'SocketServer',
+
+  documentation: `Waits on the socket connection for requests, passing them off to a SocketServerProcessor.`,
```
no need for backquote.

kgrgreer commented on src/foam/nanos/medusa/CheckWebAgent.java:
```diff
@@ -0,0 +1,61 @@
+   to add/remove targets from target group.
+   Responses from haproxy.org
+ */
+public class CheckWebAgent
```
Perhaps a more descriptive name would be helpful. 'Check' is kind of vague.

jlhughes replied:
Renamed to HealthCheckWebAgent.

jlhughes committed:
Rename CheckWebAgent to HealthCheckWebAgent

jlhughes committed:
If user is admin, allow spid override from Theme

jlhughes committed:
Replace single back-quotes with regular single quotes

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Use Theme to find spid, rather than relaying on AppConfig.

Global * spid removed as default on Theme. Now expect each Theme to have a spid.
Reenabled interface ServiceProviderAware on Theme but still accessible during
from ServiceProviderAwareDAO itself via the localThemeDAO which disables the
ServiceProviderAware interface. Also User and Group findTheme use the
localThemeDAO.

jlhughes committed:
Revert enabling ServiceProviderAware on Theme and using localThemeDAO - simply does not work.

jlhughes committed:
update doc, logging

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-157/Medusa

jlhughes committed:
Add convenience constructor to quickly create Subjects for context put calls

jlhughes committed:
move ClusterConfigSupport and ClusterConfigMonitor into mm and mn deployment journals so they are not started for non medusa builds

jlhughes committed:
Improve gap testing - to handle arbitrary gap sizes.

jlhughes committed:
add and comment out logging

jlhughes committed:
In Fallback, explicitly look for 'foam' theme

jlhughes committed:
Output both name and id of theme found with no spid
